File: rules_utils/automation/lib/python/community/rules_utils.py
"""
Copyright June 25, 2020 Richard Koshak

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

from core.rules import rule
from core.triggers import when
from core.metadata import get_value
from core.jsr223 import scope
from core.log import log_traceback


@log_traceback
def create_switch_trigger_item(item_name, logger):
    """Checks to see if the passed in Item exists and if it doesn't creates it
    as a Switch Item.

    Arguments:
        - item_name: the name of the Item to create
        - logger: used to log out informational statement
    Returns:
        True if the Item was created or it exists already and it's the right
        type. False otherwise.
    """

    if item_name not in scope.items:
        from core.items import add_item
        logger.info("Creating Item {}".format(item_name))
        add_item(item_name, item_type="Switch")
        return True

    # TODO: figure out how to test the Item's type.
#    if not isinstance(scope.ir.getItem(item_name),
#                      (scope.SwitchItem,
#                       scope.DimmerItem,
#                       scope.ColorItem)):
#        logger.error("Item {} already exists but it can't receive ON commands"
#                     .format(item_name))
#        return False
    else:
        return True


@log_traceback
def delete_rule(function, logger):
    """Deletes a rule attached to the passed in function, if it exists.

    Arguments:
        - function: the function the rule is attached to
        - logger: used to log out information and errors
    Returns:
        True if the rule is successfully deleted, False otherwise.
    """

    if hasattr(function, "UID"):
        logger.info("Deleting rule {}".format(function.UID))
        scope.scriptExtension.get("ruleRegistry").remove(function.UID)
        delattr(function, "triggers")
        delattr(function, "UID")

    if hasattr(function, "UID"):
        logger.error("Failed to delete rule {}".format(function.UID))
        return False
    else:
        return True


@log_traceback
def create_rule(name, triggers, function, logger, description=None, tags=None):
    """Creates a rule with the passed in data

    Arguments:
        - name: the name for the rule
        - triggers: a list of rule trigger strings
        - function: the function to attach the rule to
        - logger: logs out informational and error messages
        - description: optional description for the rule
        - tags: optional list of tags
    Returns:
        True if successful, False otherwise.
    """

    for trigger in triggers:
        when(trigger)(function)

    rule(name, description, tags)(function)

    if hasattr(function, "UID"):
        logger.info("Successfully created rule {}".format(name))
        return True
    else:
        logger.error("Failed to create rule {}".format(name))
        return False


@log_traceback
def create_simple_rule(item_name, name, function, logger, description=None,
                       tags=None):
    """Creates a rule triggered by a Switch Item receiving an ON command whcih
    calls the passed in function. If the Item doesn't exit, it's automatically
    created.

    Arguments:
        - item_name: Name of the Switch Item to create
        - name: Name of the rule
        - function: function to call when the rule is triggered
        - logger: logs out errors and information
        - description: description for the rule
        - tags: tags for the rule
    Returns:
        True if successful, False otherwise.
    """
    if create_switch_trigger_item(item_name, logger):
        triggers = ["Item {} received command ON".format(item_name)]
        if create_rule(name, triggers, function, logger, description, tags):
            return True
    return False


@log_traceback
def load_rule_with_metadata(namespace, check_config, event, rule_name, function,
                            logger, description=None, tags=None, systemstarted=False):
    """Creates a rule triggered by event on all Items with valid a valid
    configuration for the passed in namespace. Based additions to the original
    submission of the Expire binding to the openhab-helper-libraries by
    CrazyIvan359.


    Arguments:
        - namespace: the metadata namespace to search for.
        - check_config: a function called to check the validity of the config;
        it expects the item name as the first argument and logger as the second.
        check_config should log errors and warnings explaining what is wrong with
        the config
        - event: the event part of the rule trigger, e.g. "received command ON"
        - rule_name: The name to apply to the rule
        - function: called when the rule triggers
        - logger: log errors and information statements
        - descrption: optional description to apply to the rule
        - tags: optional tags to apply to the rule
        - systemstarted: optional system started trigger
    Returns:
        None if there was an error creating the rule, the list of Item names
        for which a trigger was created on the new rule if successful.
    """

    logger.info("Creating {} rule...".format(rule_name))

    # Remove the existing rule if it exists.
    if not delete_rule(function, logger):
        return None

    # Generate the rule triggers with the latest metadata configs.
    triggers = generate_triggers(namespace, check_config, event, logger)
    if not triggers:
        logger.warn("There are no Items with a valid {} metadata"
                    .format(namespace))
        return None

    if systemstarted:
        triggers = ["System started"] + triggers

    # Create the rule.
    if not create_rule(rule_name, triggers, function, logger, description, tags):
        return None

    return get_items_from_triggers(triggers)


def generate_triggers(namespace, check_config, event, logger):
    """Generates a trigger for all Items with namespace metadata that passes
    check_config. The event is defined in event.

    Arguments:
        - namespace: the metadata namespace to look for
        - check_config: function to call to check the validity of the metadata
        - event: rule trigger event (e.g. changed)
        - logger: used to log out errors and informational statement
    Returns:
        A list of rule trigger strings.
    """

    triggers = []
    for i in [i for i in scope.items if get_value(i, namespace)]:
        if check_config(i, logger):
            triggers.append("Item {} {}".format(i, event))

    return triggers


def get_items_from_triggers(triggers):
    """Given a list of Item rule triggers, extract the names of the Items and
    return them in a list.

    Arguments:
        - triggers: the list of rule trigger strings.
    Returns:
        A list of item_names.
    """

    return [t.split(" ")[1] for t in triggers if t.startswith("Item")]

File: rulesUtils.js
const RuleManager = osgi.getService('org.openhab.core.automation.RuleManager');
const ruleRegistry = osgi.getService('org.openhab.core.automation.RuleRegistry');
const { itemRegistry } = require('@runtime');

/**
 * Generates an Array of Item triggers for each Item in the list.
 * @param {Array<Item>} list list of Items to generate triggers for
 * @param {triggers.ItemChangedTrigger|triggers.ItemStateUpdateTrigger|triggers.ItemStateUpdateTrigger} trigger type of trigger to generate for each Item in the list
 * @param {int} systemStarted optional, when provided adds a system runlevel trigger. Valid values are 40, 50, 60, 70, 80, 90, or 100.
 * @returns {Array<triggers>} the generated rule triggers
 */
const generateTriggers = (list, trigger, systemStarted) => {
  let triggers = [];
  list.forEach(i => {
    triggers.push(trigger(i.name));
  });
  if(systemStarted && systemStarted >= 40) triggers.push(triggers.SystemStartLevelTrigger(systemStarted));
  return triggers;
}

/**
 * Checks to see if the rule exists
 * Only works for rules defined in the same file so it's utility is suspect
 * @param {String} uid UID for the rule to look for
 * @returns {Boolean} true when the rule exists
 */
const ruleExists = (uid) => {
  return !(RuleManager.getStatusInfo(uid) == null);
}

/**
 * DEPRECATED
 *
 * Deletes the rule if it exists. Does not work in UI rules.
 * @param {String} uid UID for the rule to remove
 */
const removeRule = (uid) => {
  console.warn('rulesUtils.removeRule() is deprecated, use rules.removeRule() instead.');
  if(ruleExists(uid)) {
    ruleRegistry.remove(uid);
    return !ruleExists(uid);
  }
  else {
    return false;
  }
}

/**
 * If a rule with the given UID exists, it's deleted. Then a new rule is created
 * with the given properties. A DYNAMIC_RULE_TAG will be applied to the rule.
 * @param {string} uid unique identifier for the given rule
 * @param {string} ruleName name for the rule as it will appear in MainUI
 * @param {string} ruleDescription explanation of the purpose of the rule
 * @param {Array<triggers.*>} trigs triggers which willcause the rule to run
 * @param {function(event)} func called when the rule triggers
 * @param {Array<string>} [ts=[]] optional list of tags to apply to the rule
 */
const recreateRule = (uid, ruleName, ruleDescription, trigs, func, ts = []) => {
  console.warn('rulesUtils.recreateRule() is known not to work in all cases');
  console.log('Removing the rule if necessary');
  // Remove the existing rule if it exists, only works for rules created in the same file
  if(ruleExists(uid)) if(removeRule(uid)) return null;

  console.log('Old rule removed if it existed');
  ts.push['DYNAMIC_RULE_TAG'];

  let args = {
    name: ruleName,
    description: ruleDescription,
    id: uid,
    triggers: trigs,
    tags: ts,
    execute: func
  };
  console.log('Rule args:', args);


  return rules.JSRule( {
    name: ruleName,
    description: ruleDescription,
    id: uid,
    triggers: trigs,
    tags: ts,
    execute: func
  });

}

/**
 * Creates a rule that is triggered when any Item with valid metadata in the passed in
 * namespace using the passed in trigger type. In addition to the passed in tags,
 * a DYNAMIC_RULE_TAG will also be applied.
 * @param {string} namespace Item metadata namespace used to identify those Items to trigger
 * @param {function(itemName)} checkConfig function that returns true if the metadata on the Item is valid
 * @param {triggers.ItemStateChangeTrigger|triggers.ItemStateUpdateTrigger|triggers.ItemChangeTrigger} event the Item trigger type to create for each Item
 * @param {string} ruleName name of the rule generated
 * @param {function(event)} func function to call when the rule is triggered
 * @param {string} description optional description for the rule
 * @param {Array<string>} tags optional tags to apply to the rule
 * @param {string} uid optional UID to identify the rule
 * @param {int} systemStarted optional when true a runlevel 40 trigger will be added to the rule.
 * @returns {HostRule} the created rule, or null if there was a problem
 */
const createRuleWithMetadata = (namespace, checkConfig, trigger, ruleName, func,
                                description, uid, systemStarted, tags) => {
  // Get the Items
  let triggeringItems = utils.javaSetToJsArray(itemRegistry.getAll())
       .filter(i => items.getItem(i.name).getMetadataValue(namespace) && checkConfig(i.name));

  triggers = generateTriggers(triggeringItems, trigger, systemStarted);
  if(!triggers) return null;
  return recreateRule(uid, ruleName, description, triggers, func, tags);

}

/**
 * Creates a rule that is triggered when any Item with a given tag trigger type.
 * In addition to the passed in tags, a DYNAMIC_RULE_TAG will also be applied.
 * @param {string} tag Item tag used to identify those Items to trigger
 * @param {triggers.ItemStateChangeTrigger|triggers.ItemStateUpdateTrigger|triggers.ItemChangeTrigger} event the Item trigger type to create for each Item
 * @param {string} ruleName name of the rule generated
 * @param {function(event)} func function to call when the rule is triggered
 * @param {string} description optional description for the rule
 * @param {Array<string>} tags optional tags to apply to the rule
 * @param {string} uid optional UID to identify the rule
 * @param {int} systemStarted optional when true a runlevel 40 trigger will be added to the rule.
 * @returns {HostRule} the created rule, or null if there was a problem
 */
const createRuleWithTags = (tag, trigger, ruleName, func, description,
                            uid, systemStarted, tags) => {
  let triggeringItems = items.getItemsByTag(tag);
  triggers = generateTriggers(triggeringItems, trigger, systemStarted);
  if(!triggers) return null;
  return recreateRule(uid, ruleName, description, triggers, func, tags);
}

/**
 * DEPRECATED
 *
 * Calls another rule based on UID or rule name. TODO remove when added to openhab-js
 * @param {string} nameOrUid Rule UID or name to run
 * @param {dict} argsDict optional dict of data to pass to the called rule
 * @param {boolean} cond optional flag, when true the called rule's conditions will be evaluated
 * @returns {boolean} true when the rule is found and successfully called
 */
const runRule = (nameOrUid, argsDict, cond = false) => {
  console.warn('rulesUtils.runRule() is deprecated, use rules.runRule() instead');

  // If it's not a UID, try to find it by name
  if(!RuleManager.getStatusInfo(nameOrUid)) {
    const { ruleRegistry } = require('@runtime/RuleSupport');
    const rule = utils.javaSetToJsArray(ruleRegistry.getAll())
                      .find(rule => rule.getName() == nameOrUid);
    nameOrUid = (rule) ? rule.getUID() : null;
  }

  // run the rule
  if(nameOrUid) {
    RuleManager.runNow(nameOrUid, cond, argsDict);
    return true;
  }
  else {
    return false;
  }
}

module.exports = {
  generateTriggers,
  recreateRule,
  createRuleWithMetadata,
  createRuleWithTags,
  runRule,
  ruleExists,
  removeRule
}